lib/axon/loop.ex

```diff
@@ -1547,6 +1528,118 @@ defmodule Axon.Loop do
     struct!(Axon.Loop.State, state_map)
   end
 
+  @doc """
+  Creates an Elixir `Stream` from the given loop, data, and state.
+
+  The stream will lazily return the loop state after each training epoch.
+
+  ## Options
+
+  """
+  def stream(
+        %Loop{
+          init: init_fn,
+          step: step_fn,
+          handlers: handler_fns,
+          metrics: metric_fns,
+          attached_state: attached_state
+        },
+        data,
+        init_state \\ %{},
+        opts \\ []
+      ) do
+    {max_iterations, opts} = Keyword.pop(opts, :iterations, -1)
+    {jit_compile?, opts} = Keyword.pop(opts, :jit_compile?, true)
+    {strict?, opts} = Keyword.pop(opts, :strict?, true)
+    {force_garbage_collection?, jit_opts} = Keyword.pop(opts, :force_garbage_collection?, false)
+    debug? = Keyword.get(jit_opts, :debug, false)
+
+    sample_data =
+      case Enum.take(data, 1) do
```

Comment:
This is tricky...

If the stream is pointing to an external resource, then you are opening it and traversing it. If the stream has side-effects, this first element is effectively lost. The correct way is to do this when streaming starts. Something like this:

```
Stream.unfold({:init, arg1, arg2}, fn
  {:init, arg1, arg2}, data ->
    state = init_loop_state(...)
    {emit_next(state), {:state, state}
  {:state, state}, data ->
    ....
  :halt, _data ->
    ...
end)
```

Reply:
I think you would have to move everything to `init`, including `fire_event(:started, ...)`.